Bugfixing Test/v2.1 setup process

NxFerrara opened the pull request:
## Changes:
- Fixed runtime issues with screen.py
- Fixed syntax issues with service file templates
- Fixed incorrect variable referencing in setup.sh

## Considerations:
- Do not merge until all branches to be merged into `bugfix/setup` are merged

NxFerrara committed:
Fix spaces around var assignment

The shell script interprets the variable names as commands due to the spacing around the equals sign.

NxFerrara committed:
Fix function name reference

NxFerrara committed:
Change apt -> apt-get in 3_create_venv.sh

apt-get is meant for the CLI in scripting while apt is a simplified version that is meant to be user-friendly.

NxFerrara committed:
Fix environment var name reference

NxFerrara committed:
Fix bad service unit file

The working directory cannot be in quotes otherwise it gives an error.

NxFerrara committed:
Fix empty device state file

There needs to be a default of {} in the file otherwise you cannot read the file as JSON.

NxFerrara committed:
Fix empty device state case when reading

Reading from an empty dictionary gave an error since the device state dictionary had no 'saved_timestamp' key.

NxFerrara committed:
Add popup limit for invalid taps

There is an issue with the invalid taps having many popups queued, which essentially makes the display inoperable until the queue is empty again. The popup limit aims to avoid these situations.

NxFerrara committed:
Fix time comparison error

The time was being compared lexicographically, which may have caused issues.

NxFerrara committed:
Fix variable substitution with braces

jbesoto commented on src/utils/screen_utils.py:
```diff
@@ -144,10 +144,11 @@ def read_device_state(path_to_device_state: str) -> Dict[str, Any]:
     except json.JSONDecodeError:
         logger.error('JSON Decode Error occurred reading from device state')
         raise json.JSONDecodeError # TODO: Determine error message format
```
This comment has to be removed. I believe we already established a format for the logs

NxFerrara replied:
Yup. Changes made.

NxFerrara committed:
Add time release function of SelfReleasingLock

NxFerrara committed:
Fix types, comments, and style

NxFerrara committed:
Fix global variable reference

jbesoto commented on src/utils/screen_utils.py:
```diff
@@ -144,10 +144,11 @@ def read_device_state(path_to_device_state: str) -> Dict[str, Any]:
 
-    if is_at_least_next_day(saved_timestamp, current_time):
+    if not saved_timestamp or is_at_least_next_day(saved_timestamp, current_time):
         device_state['unit_count'] = 0
```
Function name is misleading. `read_device_state` implies that the function won't change the state of the device. However, it does change it. You might want to re-think your implementation.

This is my rule of thumb: functions accomplish concepts; a concept conveyed by the function's name. The reason why I say concepts and not tasks is because having a function for each little task would make the code very hard to read. Stick to one concept per function.

NxFerrara replied:
I agree, good call. I made an attempt. Let me know what you think.

jbesoto commented on src/utils/screen_utils.py:
```diff
@@ -144,10 +144,11 @@ def read_device_state(path_to_device_state: str) -> Dict[str, Any]:
     except json.JSONDecodeError:
         logger.error('JSON Decode Error occurred reading from device state')
```
Message: "JSON Decode Error occurred...", yeah. But what does that mean? Maybe a better error message would be `Unable to parse device state: Incorrect JSON formatting`.

NxFerrara replied:
Totally, make it less cryptic. Changes made.

jbesoto reviewed (changes requested):
All good except for the Python script. Reviewing just that one code block from the `screen_utils.py` made me realize that from now on we can't allow someone to review and merge their own PRs. Let me know if you got any questions :)

NxFerrara committed:
Modularize read_device_state

The resetting of the unit count should happen independently from the reading of the device state for readability, robustness, etc.

NxFerrara committed:
Remove unnecessary comments

NxFerrara committed:
Make JSONDecodeError message clearer

NxFerrara committed:
Fix read/write device state logic

The fix was to make the code more readable, so the read_device_state function should solely read the device state, and nothing more. Same applies to write_device_state.

NxFerrara committed:
Make read/write device state handled by touch_sensor

NxFerrara committed:
Revert "Make read/write device state handled by touch_sensor"

This reverts commit 419ae203396aaf14939d3a6617054519d6e2ff50.

jbesoto commented on src/utils/screen_utils.py:
```diff
@@ -121,8 +121,25 @@ def is_at_least_next_day(from_timestamp: float, to_timestamp: float) -> bool:
     return to_day_start > from_day_start
 
 
+def check_and_reset_device_state(device_state: Dict[str, Any], path_to_device_state: str) -> None:
```
Back to square one... Is this function really necessary? If so, why? This is not what we discussed.

NxFerrara replied:
Check again, I think you reviewed old changes.

jbesoto reviewed (changes requested):
@NxFerrara still not there yet. Please implement the changes that we discussed.

jbesoto commented on scripts/run/shutdown.sh:
```diff
@@ -45,17 +45,20 @@ function load_env_variables() {
+    local target_time="$1"
+    local current_time=$(date +"%H:%M:%S")
+    
+    local target_secs=$(date -d "${target_time}" +%s)
```
Fix delete those newlines. Newline separate logical steps in the code. What are these newlines separating?

jbesoto reviewed (approved):
Good to go. Before merging make sure you adhering to the style guide and functions are properly commented.

NxFerrara committed:
Merge branch 'bugfix/setup' into refactor/nxferrara

NxFerrara committed:
Merge pull request #37 from NxFerrara/refactor/nxferrara

Refactor source files

NxFerrara committed:
Make insert_device_id() upsert

The idea of upserting is inserting or updating a record if it already exist. This needs to be the case since that is what we are doing.

nick-ferrara committed:
Merge branch 'test/v2.1' into bugfix/setup
